### PR TITLE
HDDS-8505. ReplicationManager: Add configurable global replication limit

### PR DESCRIPTION
**Pull request by sodonnel**
## What changes were proposed in this pull request?

We should make it possible to configure a global replication limit, limiting the number of inflight containers pending creation. A larger cluster would be capable of having more inflight replication than a smaller cluster, so the limit should be a function of the number of datanodes on the cluster, and the limit of the number of commands which can be queued per datanode and some weighting factor.

For example, if each datanode can queue 20 replication commands, and there are 100 nodes in the cluster, then the natural limit is 20 * 100. However, that assumes that commands are queued evenly across all datanodes, which is unlikely. With a global limit we would prefer that all datanodes are not fully loaded with replication commands simultaneously, so we may want to impose a limit of half that number, with a factor of 0.5, eg 20 * 100 * 0.5 = 1k pending replications.

At one extreme this would result in all datanodes in the cluster having half their maximum tasks queued, but in practice, some DNs are likely to be at their limit while others have zero or less than half queued.

If the limits were perfectly defined, such that in a single heartbeat a datanode can complete all its queued work just at the end of the heartbeat interval, then reducing the number of queued commands by half would make the datanode busy for only half its heartbeat interval. As the datanodes will all heartbeat at different times, all the busy and non-work periods across all the datanodes would combine in a load profile that would show some datanodes are always idle, reducing the overall load on the cluster.

This PR introduces a default limit factor of 0.75, which can be disabled by setting hte factor to 0. Only inflight replications are throttled via this mechanism - delete containers are throttled by the per datanode limits introduced some time back.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8505

## How was this patch tested?

New unit tests

### CODE REVIEW

**Review by adoroszlai — APPROVED** [review]

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/replication/TestUnderReplicatedProcessor.java [hunk 1]
```diff
@@ -93,4 +96,40 @@ public void testMessageRequeuedOnException() throws IOException {
     assertEquals(1, queue.underReplicatedQueueSize());
     assertSame(result, queue.dequeueUnderReplicatedContainer());
   }
+
+  @Test
+  public void testMessageNotProcessedIfGlobalLimitReached() throws IOException {
+    AtomicLong inFlightReplications = new AtomicLong(11);
```
Comment (swamirishi) [inline hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/replication/TestUnderReplicatedProcessor.java:102]: Can we have a case to check if replication manager is processing all the requests if inflight replication limit factor is set to 0.

Comment (sodonnel) [reply hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/replication/TestUnderReplicatedProcessor.java:102]: At line 64 in the diff, I have set the default case to be the "no limit", so if existing tests are working, then they must be OK with that. Adding any more would just be a duplicate of the existing "success" test which runs a container through successfully.

**Comment by kerneltime** [comment]
We should add metrics to be able to track this behavior for each replication scenario separately.

**Comment by adoroszlai** [comment]
Thanks @sodonnel for the patch, @swamirishi for the review.

**Pull request merged**